File: src/features/typing/data/java/largest_rectangle_in_histogram.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        int best = 0;
        for (int i = 0; i < heights.length; i++) {
            while (!stack.isEmpty() && heights[stack.peek()] >= heights[i]) {
                int h = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                best = Math.max(best, h * width);
            }
            stack.push(i);
        }
        while (!stack.isEmpty()) {
            int h = heights[stack.pop()];
            int width = stack.isEmpty() ? heights.length : heights.length - stack.peek() - 1;
            best = Math.max(best, h * width);
        }
        return best;
    }
}

File: src/features/typing/data/dart/largest_rectangle_in_histogram.dart
class Solution {
  int largestRectangleArea(List<int> heights) {
    final stack = <int>[];
    int best = 0;
    void popArea(int right) {
      final int h = heights[stack.last];
      stack.removeLast();
      final int width = stack.isEmpty ? right : right - stack.last - 1;
      best = best > h * width ? best : h * width;
    }
    for (int i = 0; i < heights.length; i++) {
      while (stack.isNotEmpty && heights[stack.last] >= heights[i]) {
        popArea(i);
      }
      stack.add(i);
    }
    while (stack.isNotEmpty) {
      popArea(heights.length);
    }
    return best;
  }
}
